indexing-hadoop/src/main/java/io/druid/indexer/HadoopDruidIndexerMapper.java

```diff
@@ -109,4 +153,37 @@ public final static InputRow parseInputRow(Object value, InputRowParser parser)
   abstract protected void innerMap(InputRow inputRow, Object value, Context context, boolean reportParseExceptions)
       throws IOException, InterruptedException;
 
+  private List<String> mergedDimensions = null;
+
+  private InputRow mergePartitionDimValues(InputRow inputRow)
+  {
+    InputRow merged = inputRow;
+
+    // only for raw data case
+    if (inputRow instanceof MapBasedInputRow) {
+      MapBasedInputRow mapBasedInputRow = (MapBasedInputRow)inputRow;
+
+      if (mergedDimensions == null) {
+        List<String> orgDimensions = mapBasedInputRow.getDimensions();
+        mergedDimensions = Lists.newArrayListWithCapacity(orgDimensions.size() + partitionDims.size());
+        mergedDimensions.addAll(orgDimensions);
+        for (String partitionDimension : partitionDims) {
+          if (!mergedDimensions.contains(partitionDimension)) {
+            mergedDimensions.add(partitionDimension);
+          }
+        }
+      }
+
+      Map<String, Object> eventMap = Maps.newHashMap(mapBasedInputRow.getEvent());
+      eventMap.putAll(partitionDimValues);
+
+      merged = new MapBasedInputRow(
+          mapBasedInputRow.getTimestamp(),
+          mergedDimensions,
+          eventMap
+      );
+    }
+
+    return merged;
+  }
```

Comment:
I guess it would be cleaner to implement an InputFormat that understands the nuances of "partitioned paths" and generates InputRows appropriately instead of having very specialized logic in this class.